astropy/coordinates/sky_coordinate.py

```diff
@@ -65,19 +66,32 @@ def _repr_data(self):
         return repr_data
 
     def _represent_as_dict(self):
-        obj = self._parent
-        attrs = (list(obj.representation_component_names) +
-                 list(frame_transform_graph.frame_attributes.keys()))
-
-        # Don't output distance if it is all unitless 1.0
-        if 'distance' in attrs and np.all(obj.distance == 1.0):
-            attrs.remove('distance')
+        sc = self._parent
+        attrs = list(sc.representation_component_names)
+
+        # Don't output distance unless it's actually distance.
+        if issubclass(sc.data.__class__, UnitSphericalRepresentation):
+            attrs = attrs[:-1]
```

Comment:
This assumes `distance` is the last attribute. Maybe someone could subclass `UnitSphericalRepresentation` and add one?

Reply:
Similarly, here just `if isinstance(sc.data, UnitSphericalRepresentation):` 

In this case, I think you are allowed to assume that someone who subclasses `UnitSpherical` will not change the order...

Reply:
Fair enough.